Mathlib/LinearAlgebra/QuadraticForm/Basic.lean

```diff
@@ -500,397 +512,437 @@ end SemiringOperators
+variable [IsScalarTower S R M] [IsScalarTower S R N]
+
+/-- If `B : M → N → Pₗ` is `R`-`S` bilinear and `R'` and `S'` are compatible scalar multiplications,
+then the restriction of scalars is a `R'`-`S'` bilinear map. -/
+@[simps!]
+def restrictScalars (Q : QuadraticMap R M N) : QuadraticMap S M N where
+  toFun x := Q x
+  toFun_smul a x := by
+    simp [map_smul_of_tower]
+  exists_companion' :=
+    let ⟨B, h⟩ := Q.exists_companion
+    ⟨B.restrictScalars₁₂ (S := R) (R' := S) (S' := S), fun x y => by
+      simp only [LinearMap.restrictScalars₁₂_apply_apply, h]⟩
+
+end restrictScalars
+
 section Comp
 
-variable [CommSemiring R] [AddCommMonoid M] [Module R M]
-variable [AddCommMonoid N] [Module R N]
+variable [CommSemiring R] [AddCommMonoid M] [Module R M] [AddCommMonoid N] [Module R N]
+variable {M' : Type v} [AddCommMonoid M'] [Module R M']
 
-/-- Compose the quadratic form with a linear function. -/
-def comp (Q : QuadraticForm R N) (f : M →ₗ[R] N) : QuadraticForm R M where
+/-- Compose the quadratic map with a linear function on the right. -/
+def comp (Q : QuadraticMap R M' N) (f : M →ₗ[R] M') : QuadraticMap R M N where
   toFun x := Q (f x)
   toFun_smul a x := by simp only [map_smul, f.map_smul]
   exists_companion' :=
     let ⟨B, h⟩ := Q.exists_companion
     ⟨B.compl₁₂ f f, fun x y => by simp_rw [f.map_add]; exact h (f x) (f y)⟩
-#align quadratic_form.comp QuadraticForm.comp
+#align quadratic_form.comp QuadraticMap.comp
 
 @[simp]
-theorem comp_apply (Q : QuadraticForm R N) (f : M →ₗ[R] N) (x : M) : (Q.comp f) x = Q (f x) :=
+theorem comp_apply (Q : QuadraticMap R M' N) (f : M →ₗ[R] M') (x : M) : (Q.comp f) x = Q (f x) :=
   rfl
-#align quadratic_form.comp_apply QuadraticForm.comp_apply
+#align quadratic_form.comp_apply QuadraticMap.comp_apply
 
-/-- Compose a quadratic form with a linear function on the left. -/
+/-- Compose a quadratic map with a linear function on the left. -/
 @[simps (config := { simpRhs := true })]
-def _root_.LinearMap.compQuadraticForm [CommSemiring S] [Algebra S R] [Module S M]
-    [IsScalarTower S R M] (f : R →ₗ[S] S) (Q : QuadraticForm R M) : QuadraticForm S M where
+def _root_.LinearMap.compQuadraticMapAux (f : N →ₗ[R] M') (Q : QuadraticMap R M N) :
+    QuadraticMap R M M' where
   toFun x := f (Q x)
-  toFun_smul b x := by simp only [Q.map_smul_of_tower b x, f.map_smul, smul_eq_mul]
+  toFun_smul b x := by simp only [map_smul, f.map_smul]
   exists_companion' :=
     let ⟨B, h⟩ := Q.exists_companion
-    ⟨(B.restrictScalars₁₂ S S).compr₂ f, fun x y => by
-      simp_rw [h, f.map_add, LinearMap.compr₂_apply, LinearMap.restrictScalars₁₂_apply_apply]⟩
-#align linear_map.comp_quadratic_form LinearMap.compQuadraticForm
+    ⟨B.compr₂ f, fun x y => by simp only [h, map_add, LinearMap.compr₂_apply]⟩
+
+/-- Compose a quadratic map with a linear function on the left. -/
+@[simps! (config := { simpRhs := true })]
+def _root_.LinearMap.compQuadraticMap [CommSemiring S] [Algebra S R] [Module S N] [Module S M]
```

Comment:
This definition now feels a little weird; I'll have another look later.

Reply:
Let's swap these around, and call this one `compQuadraticMap'`, leaving `compQuadraticMap` for the simpler one above.

Reply:
@eric-wieser do you mean `compQuadraticMapAux` should be renamed to `compQuadraticMap` and `compQuadraticMap` renamed to `compQuadraticMap'`?

Reply:
Yes

Reply:
Done.